File: 3-SAT/FormulaCreator/ITabularFormula.cs
using System.Collections.Generic;

namespace FormulaCreator
{
    public interface ITabularFormula
    {
        int VarCount { get; }
        int TermCount { get; }
        IList<IList<ITriplet>> Tiers { get; }
        bool TiersSorted { get; }
        void SortTiers();
        void Add(ITriplet triplet);
        void Add(IEnumerable<ITriplet> triplet);
        IList<int> Premutation { get; }
        void ApplyJoin(JoinInfo joinInfo, IList<ITriplet> triplets);
        bool IsEmpty { get; }
    }
}

File: 3-SAT/3-sat/3-sat-core/src/main/java/com/anjlab/sat3/ITabularFormula.java
package com.anjlab.sat3;


public interface ITabularFormula
{
    int getVarCount();
    int getTermCount();
    GenericArrayList<ITier> getTiers();
    boolean tiersSorted();
    void sortTiers();
    void add(ITriplet triplet);
    IPermutation getPermutation();
    void applyJoin(JoinInfo joinInfo, ITier tier);
    boolean isEmpty();
}
